src/init.cpp

```diff
@@ -342,6 +343,7 @@ void DynamicDataLoader::unload_data()
     SNIPPET.clear_snippets();
     vehicle_prototype::reset();
     vpart_info::reset();
+    clear_vehicle_null_part();
```

Comment:
Good job finding the cause, but the solution has a flaw: nobody outside `vehicle.cpp` should know about the variable, and especially they mustn't be obliged to reset it manually. It's an implementation detail of the module and it should remain private. IMHO there are cleaner ways (any of):
- Clear the cache inside `vpart_info::reset()`, where the source objects are cleared.
- Let `null_part` be initialized with its own static `vpart_info` and don't depend on the mutable map at all.
- Let each `vehicle` have its own static `null_part`. This way you won't need to remember to clear it, but the result shouldn't live longer than the vehicle itself, and it needlessly increases `vehicle`'s size.
- Don't  to save `info_cache` for null parts (`NULL_ID`), but it also looks like a special case.

So I'd chose either of the first two.